src/metabase/legacy_mbql/util.cljc

```diff
@@ -159,12 +159,33 @@
     [:is-null field]  [:=  field nil]
     [:not-null field] [:!= field nil]))
 
+(defn- base-type-for-legacy-field
+  [[_ _id-or-name {:keys [base-type]} :as field]]
+  (or base-type
+      ;; TODO: How to make following work in JS?
+      #?(:clj (when (integer? _id-or-name)
```

Comment:
I think this is overkill. `:base-type` should be included on every field ref these days.

This smarter desugaring of `:is-empty` is a courtesy to users who use `:is-empty` when they meant `:is-null`.

In the case where we don't already have the `:base-type` (which I'm not sure can even happen outside tests that create bad field refs), let's just preserve the old behavior and return the string-y desugaring like `[:or [:= field nil] [:= field ""]]`. Then there's no need to look up fields in the metadata.

Reply:
I did as suggested. I've also added guard in `add-type-info` string implementation, to handle un-probable case of field having no base type in options, hence being treated as `emptyable` and so empty string arg (in expansion) would be coerced to date/time type. Approach is further described in the new test.